### PR TITLE
client: check timestamp decrease and return error

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 
If timestamp decrease, the transaction consistency can break

### What is changed and how it works?
Detects timestamp decreases and returns an error when it happens.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -455,13 +459,23 @@ func (c *client) processTSORequests(stream pdpb.PD_TsoClient, requests []*tsoReq
...
+	})
 	// Server returns the highest ts.
 	logical -= int64(resp.GetCount() - 1)
 	c.finishTSORequest(requests, physical, logical, nil)
 	return nil
 }
 
 func (c *client) finishTSORequest(requests []*tsoRequest, physical, firstLogical int64, err error) {
+	if tsLessEqual(physical, firstLogical, c.lastPhysical, c.lastLogical) {
+		err = errors.Errorf("timestamp fallback, newly acquired ts (%d,%d) is less or equal to last one (%d, %d)",
```
Comment (disksing) [inline client/client.go]: can we panic here? Returning an error will be very likely to ignored because the caller usually has some retry scheme.

Comment (coocood) [reply client/client.go]: If sometimes later the new TS is greater than previous, we can go on running.

Comment (disksing) [reply client/client.go]: When this happens, it is likely that a serious problem has occurred. (Even if you don't return ts here, other clients may return ts that break the consistency)

Comment (coocood) [reply client/client.go]: ok

**Comment by coocood** [comment]
@disksing PTAL

**Comment by disksing** [comment]
Hi @coocood , there is a DATA RACE:
<details>

```
 WARNING: DATA RACE

 Write at 0x00c0010a87b0 by goroutine 43:

   github.com/pingcap/pd/client.(*client).GetTSAsync()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client.go:562 +0x1c0

   github.com/pingcap/pd/client.(*client).GetTS()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client.go:599 +0x68

   github.com/pingcap/pd/client.(*testClientSuite).TestTSO()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client_test.go:165 +0x126

   runtime.call32()

       /usr/local/go/src/runtime/asm_amd64.s:519 +0x3a

   reflect.Value.Call()

       /usr/local/go/src/reflect/value.go:308 +0xc0

   github.com/pingcap/check.(*suiteRunner).forkTest.func1()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:836 +0x9fc

   github.com/pingcap/check.(*suiteRunner).forkCall.func1()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:730 +0xb7

 

 Previous read at 0x00c0010a87b0 by goroutine 16:

   github.com/pingcap/pd/client.(*client).finishTSORequest()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client.go:481 +0x39b

   github.com/pingcap/pd/client.(*client).processTSORequests()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client.go:464 +0x91a

   github.com/pingcap/pd/client.(*client).tsLoop()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client.go:397 +0x53f

 

 Goroutine 43 (running) created at:

   github.com/pingcap/check.(*suiteRunner).forkCall()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:727 +0x4a7

   github.com/pingcap/check.(*suiteRunner).forkTest()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:818 +0x1b9

   github.com/pingcap/check.(*suiteRunner).doRun()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:659 +0x12d

   github.com/pingcap/check.(*suiteRunner).run()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:689 +0x176

   github.com/pingcap/check.Run()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/run.go:142 +0x5a

   github.com/pingcap/check.RunAll()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/run.go:105 +0xf32

   github.com/pingcap/check.TestingT()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/run.go:91 +0x770

   github.com/pingcap/pd/client.TestClient()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client_test.go:36 +0x4f

   testing.tRunner()

       /usr/local/go/src/testing/testing.go:865 +0x163

 

 Goroutine 16 (running) created at:

   github.com/pingcap/pd/client.NewClient()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client.go:170 +0x794

   github.com/pingcap/pd/client.(*testClientSuite).SetUpSuite()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client_test.go:106 +0x646

   github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

       /usr/local/go/src/fmt/scan.go:114 +0x286

   fmt.(*ss).doScanf()

       /usr/local/go/src/fmt/scan.go:1193 +0x377

   fmt.Fscanf()

       /usr/local/go/src/fmt/scan.go:143 +0xcf

   github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

       /usr/local/go/src/fmt/scan.go:114 +0x193

   github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/storage/file_storage.go:458 +0x2b3

   github.com/syndtr/goleveldb/leveldb.(*DB).checkAndCleanFiles()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/db_util.go:52 +0x2e1

   github.com/syndtr/goleveldb/leveldb.openDB()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/db.go:130 +0x9f3

   github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/storage/file_storage.go:458 +0x2b3

   fmt.Fscanf()

       /usr/local/go/src/fmt/scan.go:143 +0xcf

   github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

       /usr/local/go/src/fmt/scan.go:114 +0x193

   github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/storage/file_storage.go:458 +0x2b3

   fmt.Fscanf()

       /usr/local/go/src/fmt/scan.go:143 +0xcf

   github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

       /usr/local/go/src/fmt/scan.go:114 +0x193

   github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/storage/file_storage.go:458 +0x2b3

   fmt.Fscanf()

       /usr/local/go/src/fmt/scan.go:143 +0xcf

   github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

       /usr/local/go/src/fmt/scan.go:114 +0x286

   fmt.(*ss).doScanf()

       /usr/local/go/src/fmt/scan.go:1193 +0x377

   fmt.Fscanf()

       /usr/local/go/src/fmt/scan.go:143 +0xcf

   github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

       /usr/local/go/src/fmt/scan.go:114 +0x193

   github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/storage/file_storage.go:458 +0x2b3

   github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/db.go:476 +0xc5

   github.com/syndtr/goleveldb/leveldb.openDB()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/db.go:125 +0x9cf

   github.com/syndtr/goleveldb/leveldb.Open()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/db.go:197 +0x1e9

   fmt.Fscanf()

       /usr/local/go/src/fmt/scan.go:143 +0xcf

   github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

       /usr/local/go/src/fmt/scan.go:114 +0x193

   github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/storage/file_storage.go:458 +0x2b3

   github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/session.go:113 +0xf4

   github.com/syndtr/goleveldb/leveldb.(*session).recover()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/session.go:121 +0x18f4

   github.com/syndtr/goleveldb/leveldb.Open()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/db.go:183 +0xd7

   github.com/syndtr/goleveldb/leveldb.OpenFile()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/syndtr/goleveldb@v0.0.0-20180815032940-ae2bd5eed72d/leveldb/db.go:219 +0xa6

   github.com/pingcap/pd/server/kv.NewLeveldbKV()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/server/kv/levedb_kv.go:31 +0x52

   github.com/pingcap/pd/server/core.NewRegionStorage()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/server/core/region_storage.go:53 +0x5d

   github.com/pingcap/pd/server.(*Server).startServer()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/server/server.go:229 +0xb57

   github.com/pingcap/pd/server.(*Server).Run()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/server/server.go:302 +0xc9

   github.com/pingcap/pd/server.NewTestServer()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/server/testutil.go:47 +0xf1

   github.com/pingcap/pd/client.(*testClientSuite).SetUpSuite()

       /home/jenkins/agent/workspace/pd_test/go/src/github.com/pingcap/pd/client/client_test.go:99 +0x53

   runtime.call32()

       /usr/local/go/src/runtime/asm_amd64.s:519 +0x3a

   reflect.Value.Call()

       /usr/local/go/src/reflect/value.go:308 +0xc0

   github.com/pingcap/check.(*suiteRunner).runFixture.func1()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:785 +0x172

   github.com/pingcap/check.(*suiteRunner).forkCall.func1()

       /home/jenkins/agent/workspace/pd_test/go/pkg/mod/github.com/pingcap/check@v0.0.0-20190102082844-67f458068fc8/check.go:730 +0xb7

 ==================
```


</details>

File: client/client.go [hunk 2]
```diff
@@ -462,13 +465,26 @@ func (c *client) processTSORequests(stream pdpb.PD_TsoClient, requests []*tsoReq
 }
 
 func (c *client) finishTSORequest(requests []*tsoRequest, physical, firstLogical int64, err error) {
+	if err == nil && tsLessEqual(physical, firstLogical, c.lastPhysical, c.lastLogical) {
```
Comment (disksing) [inline client/client.go]: if err is not nil, should not update `last*` too.

**Pull request merged**